File: backend/app/schemas/transactions.py
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict


class SuggestionOut(BaseModel):
    category: str | None = None
    sub_category: str | None = None
    notes: str | None = None
    confidence: float = 0.0
    auto_prefill: bool = False


class TransactionBase(BaseModel):
    debit_date: date
    actual_date: date
    narration: str
    txn_number: str
    debit_amount: Decimal = Decimal(0)
    credit_amount: Decimal = Decimal(0)
    category: str = ""
    sub_category: str = ""
    notes: str = ""
    exclude: bool = False


class TransactionCreate(TransactionBase):
    pass


class TransactionUpdate(BaseModel):
    debit_date: date | None = None
    actual_date: date | None = None
    narration: str | None = None
    txn_number: str | None = None
    debit_amount: Decimal | None = None
    credit_amount: Decimal | None = None
    category: str | None = None
    sub_category: str | None = None
    notes: str | None = None
    exclude: bool | None = None


class TransactionOut(TransactionBase):
    id: int
    vendor_id: int | None = None
    vendor_name: str | None = None
    suggestion1: SuggestionOut | None = None
    suggestion2: SuggestionOut | None = None
    model_config = ConfigDict(from_attributes=True)


class TransactionListResponse(BaseModel):
    items: list[TransactionOut]
    total: int
    page: int
    pages: int

File: frontend/src/api/transactions.ts
import client from './client'

export interface Suggestion {
    category: string | null
    sub_category: string | null
    notes: string | null
    confidence: number
    auto_prefill: boolean
}

export interface Transaction {
    id: number
    debit_date: string
    actual_date: string
    narration: string
    txn_number: string
    debit_amount: string
    credit_amount: string
    vendor_id: number | null
    vendor_name: string | null
    category: string
    sub_category: string
    notes: string
    exclude: boolean
    suggestion1: Suggestion
    suggestion2: Suggestion
}

export interface PaginatedTransactions {
    items: Transaction[]
    total: number
    page: number
    pages: number
}

export interface TransactionUpdate {
    debit_date?: string
    actual_date?: string
    narration?: string
    txn_number?: string
    debit_amount?: string
    category?: string
    sub_category?: string
    notes?: string
    exclude?: boolean
}

export interface SplitTransactionRequest {
    splits: {
        debit_amount: number
        category: string
        sub_category: string
        notes: string
    }[]
}

export const getTransactions = (params: {
    page?: number
    page_size?: number
    search?: string
    category?: string
    vendor?: string
    start_date?: string
    end_date?: string
    min_amount?: number
    max_amount?: number
    exclude_filter?: string
}) => client.get<PaginatedTransactions>('/api/transactions', {params}).then((r) => r.data)

export const deleteTransaction = (id: number) =>
    client.delete(`/api/transactions/${id}`)

export const updateTransaction = (id: number, payload: TransactionUpdate) =>
    client.patch(`/api/transactions/${id}`, payload).then(r => r.data)

export const splitTransaction = (id: number, payload: SplitTransactionRequest) =>
    client.post(`/api/transactions/${id}/split`, payload).then(r => r.data)